### PR TITLE
CRYPTOSIM Integration to XSOAR

### PR DESCRIPTION
**Pull request by haldunerbay**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: link to the issue

## Description
A few sentences describing the overall goals of the pull request's commits.

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+            "severity": severity,
+            'type': 'Crpyotsim CorrelationAlert'
+        }
+
+        incidentsList.append(incident)
+
+
+
+    # Save the next_run as a dict with the last_fetch key to be stored
+    next_run = {'last_fetch': last_fetch}
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:160]: Since you are not increasing last_fetch, you will always be stuck at the same time

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import base64
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py:1]: Unit tests must be present all integrations commands including fetch_incidents.

**Review by bziser — COMMENTED** [review]
Good work! Here are my first comments on your code.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 1 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:160]:
```        
created_incident = datetime.strptime(time_stamp, DATE_FORMAT)
last_fetch = datetime.strptime(last_fetch,DATE_FORMAT)
if created_incident > last_fetch:
    last_fetch = created_incident.strftime(DATE_FORMAT)
```
I have creation of my alert date and compare with last_fetch. Is this well ?

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+''' INCIDENT '''
+
+
+def fetch_incidents(client: Client):
+
+    max_results = arg_to_number(arg=demisto.params().get('max_fetch'), arg_name='max_fetch', required=False)
+
+
+    first_fetch_time = arg_to_datetime(demisto.params().get('first_fetch')).isoformat()
```
[the remark this answers is not in the document]
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:108]: If it is proper, I would like to see demisto.params() in function. I prefer to don't add parameters in functions except client for now.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+DATE_FORMAT = '%Y-%m-%dT%H:%M:%SZ'  # ISO8601 format with UTC, default in XSOAR
+
+
+class Client(BaseClient):
+
+    def correlation_alerts(self):
+        args = demisto.args()
+        endTime = datetime.now()
+        startTime = endTime - timedelta(minutes=demisto.params().get('incidentFetchInterval', 360))
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:21]:
```suggestion
    def correlation_alerts(self, args, end_time=None, start_time=None):
        end_time = datetime.now() if not end_time else end_time
        start_time = end_time - timedelta(minutes=demisto.params().get('incidentFetchInterval', 360)) if not start_time else start_time
```

Comment (bziser) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:21]: We want that when you are fetching, you will pass the `start_time` by the last incident that you got the last fetch

**Review by bziser — COMMENTED** [review]
New comments:

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 4 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:21]: Okey, I changed it with end_time as parameter. And each fetch_incident trigger alerts with end_time. When end_time changed start_time is changed also because it is depending end_time

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+    max_results = arg_to_number(arg=demisto.params().get('max_fetch'), arg_name='max_fetch', required=False)
+
+
+    first_fetch_time = arg_to_datetime(demisto.params().get('first_fetch')).isoformat()
+
+    last_run = demisto.getLastRun()
+    last_fetch = last_run.get('last_fetch', first_fetch_time)
+
+    incidentsList=[]
+    alert_response = client.correlation_alerts(end_time=last_fetch)
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:111]: But the `last_fetch` is your new `start_time` not `end_time`

Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:111]: Yes, 
In this case:
start_time is depends end_time as it comes from definition.
end_time is current date or given last_fetch time. Then, start_time is determined end_time - FetchIntervalTime.
Example:
fetch_time = "2022-03-16T16:55:00"
intervalFetch = "1 hour"
if fetch_incident trigger the correlation_alerts:
end_time will be fetch_time
and start time will be  "2022-03-16T15:55:00". So start time depends the last_fetch time

Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:111]: > We want that when you are fetching, you will pass the start_time by the last incident that you got the last fetch

I changed end time is always datetime.now. but start_time is depend last_fetch and fetch interval.
```
def correlation_alerts(self, start_time=None):
    args = demisto.args()
    
    end_time = datetime.now()
    
    interval_time = end_time - timedelta(minutes=demisto.params().get('incidentFetchInterval', 360))
    
    if start_time is None or start_time < interval_time:
        start_time = interval_time
```

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+    Raises exceptions if something goes wrong.
+    :type client: ``Client``
+    :param Client: client to use
+    :return: 'ok' if test passed, anything else will fail the test.
+    :rtype: ``str``
+    """
+
+    message: str = ''
+    try:
+        if client.correlations().get('StatusCode') == 200:
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:87]: It's better to give some limit here to the function and not return all the correlations.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+            "severity": severity,
+            'type': 'Crpyotsim CorrelationAlert'
+        }
+
+        incidentsList.append(incident)
+
+        created_incident = datetime.strptime(time_stamp, DATE_FORMAT)
+        last_fetch = datetime.strptime(last_fetch, DATE_FORMAT)
+        if created_incident > last_fetch:
+            last_fetch = created_incident
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:156]: Duplicates must be avoided, so take the time of the next incident which you have not yet entered into the list.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+    try:
+        client = get_client(params)
+
+        if demisto.command() == 'test-module':
+            # This is the call made when pressing the integration Test button.
+            result = test_module(client)
+            return_results(result)
+
+        elif demisto.command() == 'cryptosim-get-correlations':
+            return_results(correlations_command(client))
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:211]: you don't want it to get some limit?

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+        elif demisto.command() == 'cryptosim-get-correlation-alerts':
+            return_results(correlation_alerts_command(client))
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:214]: You don't want the user to get the option to choose start-date and end-date?

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 6 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:87]: For this version, there is no connection test end point. This is temporary function to check can xsoar connect cryptosim or not. after development, we will update this.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 9 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:214]: For this version, correlation-alert fetch only alerts by given fetch interval. It is not optional, but we will do it. There is urgency to integration for our client who wait it.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 7 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:156]: Each alert is unique and must be converted to incident so we cannot control with time. I don't understand what you mean for that.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+''' CONSTANTS '''
+
+DATE_FORMAT = '%Y-%m-%dT%H:%M:%SZ'  # ISO8601 format with UTC, default in XSOAR
+
+
+class Client(BaseClient):
+
+    def correlation_alerts(self, start_time=None):
+        args = demisto.args()
+<<<<<<< HEAD
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:20]: what is this?

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+def fetch_incidents(client: Client):
+
+    max_results = arg_to_number(arg=demisto.params().get('max_fetch'), arg_name='max_fetch', required=False)
+
+    first_fetch_time = arg_to_datetime(demisto.params().get('first_fetch')).isoformat()
+
+    last_run = demisto.getLastRun()
+    last_fetch = last_run.get('last_fetch', first_fetch_time)
+
+<<<<<<< HEAD
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:125]: Why did you add this?

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 10 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:20]: vsc merge problem. I solved.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 7 above)
Comment (bziser) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:159]: for example: 
start = 15:00
max_fetch = 20
end = now = 17:00
We got from the API 25 alerts.
So when we will get to the 20th we are getting its `created_incident` to the `last_fetch`.
So the next fetch we will get again this 20th alert, instead of the 21th

Comment (bziser) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:159]: I hope it is much clearer, if not we can zoom

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 9 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:217]: The purpose of correlations end point to get all active correlations. There is no filter or parametric results.
For correlation alerts command, client can run command with parameters, only fetching correlation alerts function has not optional for start and end date. It fetchs according to fetch interval value given by client.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 6 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:88]: Yes for now there is no limitations for correlations function. It gives all active correlations, but for next version we will have connection_test end point and limitations for commands

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Category_ID.json [hunk 12]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (tomer-pan) [inline Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Category_ID.json:1]: @haldunerbay Why not using the existing "Category" field instead?

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Correlation_ID.json [hunk 13]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (tomer-pan) [inline Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Correlation_ID.json:1]: @haldunerbay please add a prefix to the cliname, id and name of the field. Also, please associate it with the cryptosim incident type and not with all of the types.

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Event_Start_Date.json [hunk 14]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (tomer-pan) [inline Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Event_Start_Date.json:1]: @haldunerbay Why not use the existing "Start Time" field?

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Category_ID.json (hunk 12 above)
Comment (haldunerbay) [reply Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Category_ID.json:1]: Category ID is specific key for cryptosim alerts, so category mean more general name.

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Log_Date.json [hunk 15]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (tomer-pan) [inline Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Log_Date.json:1]: @haldunerbay please add a prefix to the cliname, id and name of the field. Also, please associate it with the cryptosim incident type and not with all of the types.

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Log_Type.json [hunk 16]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (tomer-pan) [inline Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Log_Type.json:1]: @haldunerbay please add a prefix to the cliname, id and name of the field. Also, please associate it with the cryptosim incident type and not with all of the types.

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Plugin_ID.json [hunk 17]
```diff
@@ -0,0 +1,30 @@
+{
```
Comment (tomer-pan) [inline Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Plugin_ID.json:1]: @haldunerbay please add a prefix to the cliname, id and name of the field. Also, please associate it with the cryptosim incident type and not with all of the types.

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Correlation_ID.json (hunk 13 above)
Comment (haldunerbay) [reply Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Correlation_ID.json:1]: do you mean prefix is cryptosim for fields ?

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Category_ID.json (hunk 12 above)
Comment (tomer-pan) [reply Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Category_ID.json:1]: If you think it's critical, you can add it, but please add a prefix to the cliname, id and name of the field and also associate it only to the relevant incident type.

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Correlation_ID.json (hunk 13 above)
Comment (tomer-pan) [reply Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Correlation_ID.json:1]: Yes, you should add "cryptosim" as the prefix. For example, here:
cliname: "cryptosimcorrelationid"
id: "incident_cryptosimcorrelationid"
name: "Cryptosim Correlation ID"

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Category_ID.json (hunk 12 above)
Comment (haldunerbay) [reply Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Category_ID.json:1]: Added.

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Correlation_ID.json (hunk 13 above)
Comment (haldunerbay) [reply Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Correlation_ID.json:1]: Added.

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Log_Date.json (hunk 15 above)
Comment (haldunerbay) [reply Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Log_Date.json:1]: Added.

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Log_Type.json (hunk 16 above)
Comment (haldunerbay) [reply Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Log_Type.json:1]: Added.

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Plugin_ID.json (hunk 17 above)
Comment (haldunerbay) [reply Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Plugin_ID.json:1]: Added.

File: Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Event_Start_Date.json (hunk 14 above)
Comment (haldunerbay) [reply Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Event_Start_Date.json:1]: Same reason like category id. To specify event start using event word.

Comment (tomer-pan) [reply Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Event_Start_Date.json:1]: Again, preferably, you would use the existing field If you find this as an important addition, you would need to add the prefix and associate it only to the relevant incident type.

Comment (haldunerbay) [reply Packs/Cryptosim/IncidentFields/incidentfield-Cryptosim_Event_Start_Date.json:1]: I added prefixes and all of them associated with my incident type 'Cryptosim Correlation Alert'.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+                         "severity": 2,
+                         'type': 'Crpyotsim CorrelationAlert'
+                     }])
+
+
+    requests_mock.post("https://test.com/api/service/correlationalerts",
+                        json=mock_response)
+
+    next_run, incidents = fetch_incidents(test_client)
+    assert next_run == {"lastRun": "2018-10-24T14:13:20+00:00"}
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py:79]: Why the new next_run and old _next_run are the same?

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py [hunk 19]
```diff
@@ -0,0 +1,251 @@
...
+
+        if demisto.command() == 'test-module':
+            # This is the call made when pressing the integration Test button.
+            result = test_module(client)
+            return_results(result)
+
+        elif demisto.command() == 'cryptosim-get-correlations':
+            return_results(correlations_command(client))
+
+        elif demisto.command() == 'cryptosim-get-correlation-alerts':
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:230]: I see you added to "get-correlations" limit. Is it not possible to add also to "get-correlation-alerts"?

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+    requests_mock.post("https://test.com/api/service/correlationalerts",
+                        json=mock_response)
+
+
+    results = correlation_alerts_command(test_client)
+    assert results.outputs.keys() == ['StatusCode', 'Data', 'OutParameters']
+    assert results.outputs_prefix == 'CorrelationAlerts'
+
+
+def test_fetch_incidents(client):
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py:64]: For checking fetch you need to add more tests.
For example: fetch with max_fetch=5 and the fetch returned 10 items. 
So you need to see you are not missing and getting the expected result

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 19 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:230]: I added both of them. We will update end points, now limit can be given as argument.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+                         'rawJSON': json.dumps({"NAME": "test1", "id": "123"}),
+                         "severity": 2,
+                         'type': 'Crpyotsim CorrelationAlert'
+                     }])
+
+
+    requests_mock.post("https://test.com/api/service/correlationalerts",
+                        json=mock_response)
+    max_fetch = 3
+    next_run, incidents = fetch_incidents(test_client, {"max_fetch":max_fetch,"first_fetch":"2022-03-25T00:00:00"})
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py:78]: `first_fetch` is a configurable Parameter in the integration settings not a timestamp
```
- display: First Fetch Time (default 1 hour)
  defaultvalue: 1 hour
  hidden: false
  name: first_fetch
  required: false
  type: 0
```

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py [hunk 22]
```diff
@@ -0,0 +1,266 @@
...
+            "isJsonLog": True
+        }
+
+        return self._http_request("POST", url_suffix="correlationalertswithlogs",
+                                  data=json.dumps(parameters))
+
+    def correlations(self):
+        args = demisto.args()
+
+        limit = str(args.get("limit",100))
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:53]: limit is str, 
so you need to use `arg_to_number` for converting it to int, or to set 100 to string default

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py [hunk 23]
```diff
@@ -0,0 +1,266 @@
...
+    alert_response = client.correlation_alerts(last_fetch_time=last_fetch)
+    incident_data = alert_response.get("Data",[])
+
+    for inc in incident_data:
+
+        if len(incidentsList) > max_results:
+            break
+
+        incident_name = inc['CorrelationAlert']['NAME']
+        time_stamp = inc['CorrelationAlert']['CREATEDATE'] + "Z"
```
Comment (bziser) [inline Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:153]: Didn't you say you are not getting it timezoned?

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 22 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:53]: so,
limit = args.get("limit","100")
is okey for this case ?

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py (hunk 20 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py]: I wrote the max_fecth control.
    assert len(incidents) == max_fetch

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 23 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:153]: 'CREATEDATE' is local date, 
I create a instance params get time_zone and I added it to XSOAR's time values

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 22 above)
Comment (bziser) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:53]: `str(args.get("limit", '100'))`

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py (hunk 21 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py:78]: Yes, I noticed and changed as  '1 hour'

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 23 above)
Comment (bziser) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:153]: I don't understand. 
If it is local date, so you need to parse it, not only add to it "Z"

Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:153]: No no, I don't change my date, I changed the XSOAR date,
For example:
CREATEDATE: 2022-03-28T:14:00:00 server in Istanbul.
But at this time XSOAR time is 2022-03-28T:11:00:00 in UTC.
So when I compare the dates I add XSOAR time the utc difference. For difference I added instance parameter.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 22 above)
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:53]: I fixed it.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py (hunk 23 above)
Comment (bziser) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:161]: Why not to change your time to UTC from start when you get it?

Comment (bziser) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:161]: But you are not changing XSOAR time to UTC, you are changing your time to UTC (in not a good way):
`time_stamp = inc['CorrelationAlert']['CREATEDATE'] + "Z"`

Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim.py:161]: No, inc['CorrelationAlert']['CREATEDATE'] give local time but not completly iso_formatted. I added 'Z'  to tag its' utc times but not utc+0.
if time_stamp is 16:00:00 for my product , this mean is 13:00:00 for utc.
So, I added `timedelta(hours=int(demisto.params().get("time_zone_difference",3))` to XSOAR times.
time_zone_difference is taken from instance params.

File: Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+                         'type': 'Crpyotsim CorrelationAlert'
+                     }])
+
+
+    requests_mock.post("https://test.com/api/service/correlationalerts",
+                        json=mock_response)
+    max_fetch = 3
+    next_run, incidents = fetch_incidents(test_client, {"max_fetch":max_fetch,"first_fetch":"1 hour"})
+    assert next_run["lastRun"] == mock_response[0]["lastRun"]
+    assert len(incidents) == max_fetch
```
[the remark this answers is not in the document]
Comment (haldunerbay) [reply Packs/Cryptosim/Integrations/Cryptosim/Cryptosim_test.py:80]: How I got 1 incident ? In incidents list, there are incidents that limited by alert count or max_fetch.
actually line 80 should be `assert len(incidents) <= max_fetch`
because not always alert count reach the max_fetch count

**Review by bziser — APPROVED** [review]
